### DISCUSSION TITLE
use of "export" in brew.env

### DISCUSSION THREAD

**Discussion by getawaywithrmdir** [Homebrew/discussions#5232] [Everyday usage]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.2.15-10-ga570e4d
ORIGIN: https://github.com/Homebrew/brew
HEAD: a570e4da19131eb9e7263f392e863cbaf3325884
Last commit: 18 hours ago
Core tap JSON: 26 Mar 15:43 UTC
Core cask tap JSON: 26 Mar 15:43 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
HOMEBREW_NO_ENV_HINTS: set
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.1.4 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 14.4-arm64
CLT: 15.3.0.0.1.1708646388
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

I've been tearing my hair out thinking that Homebrew wasn't finding my brew.env file, but eventually I sorted out through experimentation that it doesn't like it when I use "export" before setting an environment variable--despite that being how all the examples in the manpages are shown.

My question: is this the intended behavior? If so, I think the manpages need updating; if not, then I think something else is wrong (or at least inconsistent).

For comparison, the text in my brew.env file used for that `brew config` output was:
```
export HOMEBREW_AUTOREMOVE=1
export HOMEBREW_BAT=1
HOMEBREW_NO_ENV_HINTS=1
```
_Note the presence of HOMEBREW_NO_ENV_HINTS and absence of HOMEBREW_AUTOREMOVE and HOMEBREW_BAT. Removing "export" causes them to correctly show up in `brew config` output._

In case it's somehow relevant, my config path is set by XDG_CONFIG_HOME and does not match the default MacOS location.

**Comment by SMillerDev** [comment]
What reads the brew.env file? Because as far as I know that's not something Homebrew does. All the examples are for shell init files like `~/.bashrc`

**Comment by getawaywithrmdir** [comment]
I don't source it anywhere (only occurrences of "brew" in any of my zsh files are `eval "$(/opt/homebrew/bin/brew shellenv)"` in my .zprofile and sourcing the spaceship.zsh theme, which I installed with home-brew). Per the man page:
```
HOMEBREW_* environment variables can also be set in
Homebrew’s environment files:

•   /etc/homebrew/brew.env (system-wide)

•   $HOMEBREW_PREFIX/etc/homebrew/brew.env
           (prefix-specific)

•   $XDG_CONFIG_HOME/homebrew/brew.env if
           $XDG_CONFIG_HOME is set or $HOME/.homebrew/brew.env
           otherwise (user-specific)
```
It will specifically auto-detect settings in the above files, which is how it is able to read my file, so that's working. The man pages never reference dotfiles other than to add Homebrew to your path, so I took that to mean that brew.env is the "preferred" place to organize those settings, and if it's the preferred place I would expect brew.env to work like the examples. Even if it's not "preferred," it's weird for it to mention only one specific place to put these settings while only providing examples that don't work in that one place and there's no mention of the needed adaptation to use this auto-detection feature.

There is a reference to not supporting command execution, but that's completely unclear since it says it right after saying variable expansion `$HOME` isn't supported and it only gives the example `$(cat file)`, implying it's the $ expansion syntax that's not supported--plus plenty of folks aren't going to quickly register "oh yeah, export is a command..." And it's super easy to miss for anybody just searching the man pages for the bits they need instead of reading it start to finish--just not great usability.

So, sounds like the man pages could use updating. I would love to see it called out to omit the "export" when using the suggested file locations (or to have brew honor my exports, but I bet fixing the man page is less work). I wasted so much time thinking it was not registering my XDG_CONFIG_HOME variable.

**Comment by Bo98** [comment]
> My question: is this the intended behavior?

Unlike shell init files, `brew.env` is not a shell script. It's basically a key-value based file, so there's no keywords like "export" and doesn't have shell-like parsing, e.g. quotation marks are taken literally. It will also only allow you to set `HOMEBREW_` variables - you can't set `PATH` etc here.

> The man pages never reference dotfiles

We document all the Homebrew-specific environment variables, but there's many different ways you can set this (including `/etc/environment` etc) so we don't give step-by-step instructions on how to set them. `brew.env` is however Homebrew-specific so we document that.

How you set environment variables also totally depends on where you're invoking brew from. Not every shell uses `.bashrc/.zshrc`. Not every shell even uses `export` syntax. Some people may only need environment variables temporarily for one command, some may need it to persist. Some may be calling `brew` from a programming language, which is again another syntax.

Generally: Homebrew manpage documents Homebrew but doesn't document other parts of your system such as shell init files. Setting `brew shellenv` is a "first-time setup tutorial" that somewhat stretches beyond this.

> I took that to mean that brew.env is the "preferred" place to organize those settings, and if it's the preferred place I would expect brew.env to work like the examples

The main use case of the files is system-managed environment variables. It's a fairly new feature so hasn't seen much use on user-level yet, particularly given that for most users it doesn't do much that can't already be done via existing techniques such as shell init files. If you already are doing `brew shellenv` in a shell init file then most people have kept on just placing environment variables around that.

The main advantage of `brew.env` comes when used outside of terminals (e.g. if you ran it in a daemon/cron). That doesn't matter for a lot of people but it will for some.

We don't really officially promote a "preferred" way. Given `brew.env` works more universally, it's a good option and it is a nice place so keep your settings all together, as long as you don't need to set environment variables dynamically. Homebrew will always accept environment variables so there'll also never be a requirement to use `brew.env` either. It's simply a case of choosing what you prefer - I generally prefer `brew.env` but some people may disagree.

> sounds like the man pages could use updating

Given it's a fairly new feature, the documentation of it is one of the newest parts of the manpage and as a result hasn't gained too much feedback yet - we do welcome contributions to clarify documentation.
